src/gdal/ogr_file_format.cpp

```diff
@@ -1405,5 +1379,573 @@ LatLon OgrFileImport::calcAverageLatLon(OGRDataSourceH data_source)
+		{
+			for (const auto& coord : part.path_coords)
+			{
+				QPointF proj_cord = georef.toProjectedCoords(coord.pos);
+				OGR_G_AddPoint_2D(cur_ring.get(), proj_cord.x(), proj_cord.y());
+			}
+			OGR_G_CloseRings(cur_ring.get());
+			if (need_wgs1984)
+				OGR_G_Transform(cur_ring.get(), transformation.get());
+			OGR_G_AddGeometry(polygon.get(), cur_ring.get());
+			cur_ring.reset(OGR_G_CreateGeometry(wkbLinearRing));
+		}
+
+		OGR_F_SetGeometry(po_feature.get(), polygon.get());
+
+		OGR_F_SetStyleString(po_feature.get(),
+		                     OGR_STBL_Find(table.get(), QString::number(map->findSymbolIndex(symbol)).toLatin1().constData()));
+
+		if (OGR_L_CreateFeature(layer, po_feature.get()) != OGRERR_NONE)
+			throw FileFormatException(tr("Failed to create feature in layer"));
+	};
+
+	map->applyOnMatchingObjects(add_feature, condition);
+}
+
+OGRLayerH OgrFileExport::createLayer(const char* layer_name, OGRwkbGeometryType type)
+{
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,0,0)
+	auto po_layer = GDALDatasetCreateLayer(po_ds.get(), layer_name, map_srs.get(), type, nullptr);
+#else
+	auto po_layer = OGR_DS_CreateLayer(po_ds.get(), layer_name, map_srs.get(), type, nullptr);
+#endif
+	if (!po_layer)
+		return nullptr;
+
+	if (OGR_L_CreateField(po_layer, o_name_field.get(), 1) != OGRERR_NONE)
+		addWarning(tr("Failed to create name field"));
+
+	return po_layer;
+}
+
+void OgrFileExport::populateStyleTable()
+{
+	table = ogr::unique_styletable(OGR_STBL_Create());
+	ogr::unique_stylemanager manager = ogr::unique_stylemanager(OGR_SM_Create(table.get()));
+
+	auto get_pen_style = [](const Symbol* symbol) {
+		auto main_color = symbol->guessDominantColor();
+		if (!main_color)
+			return QString::fromLatin1("PEN(c:#00000000)");
+
+		QString rgb = QColor(main_color->getRgb()).name();
+		return QString::fromLatin1("PEN(c:%1)").arg(rgb);
+	};
+
+	auto get_brush_style = [](const Symbol* symbol) {
+		auto main_color = symbol->guessDominantColor();
+		if (!main_color)
+			return QString::fromLatin1("BRUSH(fc:#00000000)");
+
+		QString rgb = QColor(main_color->getRgb()).name();
+		return QString::fromLatin1("BRUSH(fc:%1)").arg(rgb);
+	};
+
+	// Go through all used symbols and create a style table
+	std::vector<bool> symbols_in_use;
+	map->determineSymbolsInUse(symbols_in_use);
+
+	for (auto i = map->getNumSymbols() - 1; i >= 0; --i)
+	{
+		if (symbols_in_use[i])
+		{
+			const auto& symbol = map->getSymbol(i);
+			switch (symbol->getType())
+			{
+			case Symbol::Text:
+				{
+					QString new_style = QString::fromLatin1("LABEL(f:\"")
+					                    .append(symbol->asText()->getFontFamily())
+					                    .append(QString::fromLatin1("\", s:"))
+					                    .append(QString::number(12))
+					                    .append(QString::fromLatin1(", t:{Name})"));
+					OGR_SM_AddStyle(manager.get(),
+					                QString::number(map->findSymbolIndex(symbol)).toLatin1(), new_style.toLatin1());
```

Comment:
Number and new_style could be done with `QByteArray` directly. For efficiency, reserve() a reasonable size before appending. And I guess you could directly use `i` instead of `map->findSymbolIndex(symbol)`.
(Similar for the other symbol types below.)